rules/prefer-string-starts-ends-with.js

```diff
@@ -64,33 +70,62 @@ const create = context => {
 				return;
 			}
 
+			function fix(fixer, {useNullishCoalescing, useOptionalChaining, useStringCasting} = {}) {
+				const method = result.messageId === MESSAGE_STARTS_WITH ? 'startsWith' : 'endsWith';
+				const [target] = node.arguments;
+				let targetString = sourceCode.getText(target);
+				const isRegexParenthesized = isParenthesized(regexNode, sourceCode);
+				const isTargetParenthesized = isParenthesized(target, sourceCode);
+
+				if (
+					// If regex is parenthesized, we can use it, so we don't need add again
+					!isRegexParenthesized &&
+					(isTargetParenthesized || shouldAddParenthesesToMemberExpressionObject(target, sourceCode))
+				) {
+					targetString = `(${targetString})`;
+				}
+
+				if (useNullishCoalescing) {
+					// (target ?? '').startsWith(pattern)
+					targetString = (isRegexParenthesized ? '' : '(') + targetString + ' ?? \'\'' + (isRegexParenthesized ? '' : ')');
+				} else if (useStringCasting) {
+					// String(target).startsWith(pattern)
+					const isTargetStringParenthesized = targetString.startsWith('(');
```

Comment:
This check needs work.

Reply:
Let's use `getParenthesizedText` here, so this case `.test( (0, undefined) )`, we don't need add check when move to `String( (0, undefined) )`

Reply:
I significantly refactored and improved the parenthesis logic.

Reply:
I'm really sorry that I didn't make it clear  I  don't mean that one , should use [this one ](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/0b980dd719a9fdf6d3e62db17a9259190d3a30da/rules/utils/parentheses.js#L70). The one in `prefer-ternary` adding parens only fit to fix into ternary. We need write a new one here .

Reply:
Updated to use:

```
const {getParenthesizedText} = require('./utils/parentheses');
```

Reply:
I'm sent a PR to update `getParenthesizedText` in `prefer-ternary`, #1280

You need do similar thing here, 

````js
let text = getParenthesizedText(target, ...)
if (useNullishCoalescing) {
	if (
		!isParenthesized(target) &&
		shouldAddParenthesesToLogicalExpressionLeft(target)
	) 
		text = `(${text})`
	}

	// then we need use a fake LogicalExpression to check `shouldAddParenthesesToMemberExpressionObject`
	// Or just add parens to `${text} ?? ''` since it's always needed.
}

Reply:
Will try...

Reply:
I added a new function `shouldAddParenthesesToLogicalExpressionLeft` which is a copy of `shouldAddParenthesesToConditionalExpressionChild`. Not entirely sure if it should be different at all...

Reply:
Should this function be `shouldAddParenthesesToLogicalExpressionLeft` or `shouldAddParenthesesToLogicalExpressionChild`? Trying to figure out if the left side has different needs...

Reply:
I think left side is different from the right side, eg `a ?? b` to be `??`'s left don't need but if it's right, it's needed. We can still call that function `...Node` but only implement `left` part logic, because we only need that now, we also need pass operator so we should call it in this way `shouldAddParenthesesToLogicalExpressionChild(node, {operator:"??", property:"left"})`.

Reply:
> Not entirely sure if it should be different at all...

They are total different, you should consider all nodes has lower precedence according to [ this rable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#Table).